runtime/include/tt/runtime/detail/ttmetal.h

```diff
@@ -159,6 +159,19 @@ createBufferFromTensorRef(::tt::tt_metal::Device *device,
   std::shared_ptr<::tt::tt_metal::Buffer> buffer =
       ::tt::tt_metal::CreateBuffer(shardedBufferConfig);
   assert(tensorRef->address());
+
+  // Issue #408: Temporary Hack, remove when fix available.
+  // Update tt-metal BUFFER_MAP with updated address and remove
+  // entry for original alloc'd address.
+  auto &buffer_map = tt::tt_metal::detail::BUFFER_MAP;
+  auto map_copy = buffer_map.value();
+  auto old_key = std::make_tuple(device->id(), buffer->address());
+  if (auto it = map_copy.find(old_key); it != map_copy.end()) {
+    auto new_key = std::make_tuple(device->id(), tensorRef->address());
+    buffer_map.insert(new_key, it->second);
+    buffer_map.erase(old_key);
+  }
```

Comment:
Can you do all of this only `if (tensorRef->address() != buffer->address())`? 
Also, is there a reason to have a map copy?

Reply:
@rpavlovicTT , it's pretty much guaranteed that `tensorRef->address() != buffer->address()`.  This is a hack that reaches into tt-metal data structure to fix things up and avoid a segfault.  We have a proper fix in metal in flight and ideally lands in a week, but depending on metal reviewers could take longer.  We just wanted something that could unblock our silicon CI.  We'll remove this whole block once the fix in metal lands and we do another metal uplift.

Reply:
Yep, temporary hack, and yes it's guaranteed `tensorRef->address() != buffer->address()` today, the buffer->address() picked by metal will be completely unrelated, which is why it's later (after this hack) set to be `tensorRef->address()`.

For the question about map copy, due to how `tt::tt_metal::detail::BUFFER_MAP` is implemented, it only returns a copy of it's map and so if I did something like this (without storing a single copy of the map) :

`if (auto it = buffer_map.value().find(old_key); it != buffer_map.value().end())`

it doesn't compile:

```
/localdev/kmabee/mlir2/runtime/include/tt/runtime/detail/ttmetal.h:197:19: error: object backing the pointer will be destroyed at the end of the full-expression [-Werror,-Wdangling-gsl]
  197 |     if (auto it = buffer_map.value().find(old_key); it != buffer_map.value().end()) {
      |                   ^~~~~~~~~~~~~~~~~~
```